core/src/main/java/org/apache/iceberg/BaseRowDelta.java

```diff
@@ -81,23 +82,32 @@ public RowDelta validateDataFilesExist(Iterable<? extends CharSequence> referenc
   }
 
   @Override
-  public RowDelta validateNoConflictingAppends(Expression newConflictDetectionFilter) {
+  public RowDelta validateNoConflictingOperations(Expression newConflictDetectionFilter) {
     Preconditions.checkArgument(newConflictDetectionFilter != null, "Conflict detection filter cannot be null");
     this.conflictDetectionFilter = newConflictDetectionFilter;
     return this;
   }
 
+  @Override
+  public RowDelta validateNoConflictingDeleteFiles() {
+    this.validateNoConflictingDeleteFiles = true;
+    return this;
+  }
+
   @Override
   protected void validate(TableMetadata base) {
     if (base.currentSnapshot() != null) {
       if (!referencedDataFiles.isEmpty()) {
         validateDataFilesExist(base, startingSnapshotId, referencedDataFiles, !validateDeletes);
       }
 
-      // TODO: does this need to check new delete files?
       if (conflictDetectionFilter != null) {
         validateAddedDataFiles(base, startingSnapshotId, conflictDetectionFilter, caseSensitive);
       }
+
+      if (conflictDetectionFilter != null && validateNoConflictingDeleteFiles) {
+        validateAddedDeleteFiles(base, startingSnapshotId, conflictDetectionFilter, caseSensitive);
```

Comment:
I won't be able to use `validateNoConflictingAppends` here as `RowDelta` works with positions only.
What about using `DataFile` in `validateDataFilesExist`, @rdblue? Any particular reasons for using paths only?

Reply:
I don't think we will be able to use `DataFile` for `RowDelta` that easily as we are using a regular scan and delete writers just give us referenced data file locations. Then I'd say we probably still need the new method here.

Reply:
If I understand correctly, the motivation for updating `RowDelta` is the case where we have two concurrent delta commits? So an UPDATE and a MERGE at the same time might both rewrite a row, which could cause a duplicate:

```sql
INSERT INTO t VALUES (1, 'a'), (2, 'b'), (3, 'c');

-- running these concurrently causes a problem
UPDATE t SET data = 'x' WHERE id = 1;
UPDATE t SET data = 'y' WHERE id = 1;
```

If I ran the updates concurrently, both would delete id=1 and both would add a new file with `(1, 'x')` and `(1, 'y')` right?

The validation here is that the file created by the initial insert doesn't have any new delete files written against it. It seems like we want to just call `validateNoNewDeletesForDataFiles` and pass `referencedFiles` in, right? Maybe I'm missing something?

We might want to make this a separate issue to keep changes smaller and reviews easier.

Reply:
We only know locations of referenced files. We need actual `DataFile` objects to leverage the delete file index. My question is why we went with locations only in `RowDelta` and whether we should switch to `DataFile` objects instead.

Reply:
Well, I think we could approach it differently. Let me update and then we can discuss more.

Reply:
The reason why we only have file locations in `RowDelta` is that we get the set of referenced files from writing position deletes, which are `(location, position)`. We don't require carrying `DataFile` through delta operations.

Reply:
In fact, if we create a `DataFile`, we can copy the column stats from the `DeleteFile` so that we can do stats overlap comparisons.

Reply:
I had a chance to think more about the required validation. While using `DataFile`s instead of locations would give us min/max filtering, we can probably do better than that if all delete files are position-based.

We can do something like this if we are working with position-based delete files:
- Read all position deletes we are trying to commit and build a map with file location -> a list of deleted positions.
- Iterate through concurrently added delete files applying min/max filtering (secondary indexes in the future) and find which may have conflicts.
- Read files that potentially conflict and verify that (file, pos) pairs don't overlap.

We may need to add some limit on the overall size of deletes we can scan but we can figure that out.
Overall, this will allow us to resolve conflicts within the same partition.